<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера</b>


Привет, Сергей! Давай знакомиться! Меня зовут Дмитрий Махортов, и я буду проверять твой проект. Сразу предлагаю общение на «ты» 🙂, но если тебе это не комфортно, то дай знать, и мы перейдем на «вы». 


Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе погрузиться в увлекательный мир работы с данными и вырасти в крепкого профи. Это отдаленная цель. А ближайшая - сделать твою работу еще лучше )).
   
    
Все ключевые этапы в работе выполнены, и я вижу что с проектом ты справшяешься. Есть моменты, которые нужно доработать, но я уверен, у тебя все получится.
        
Расскажу как обычно проходит проверка:    
Бывают моменты, которые требуют пристального внимания. Комментарии по ним выделены <span style='background-color:#F7B3A4'> красным цветом </span> и обозначены значком 🛑. После их доработки проект будет принят.  🙂
 
<span style='background-color:#B7EBA7'> Зеленым цветом </span> и значком ✅ отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. Или советы «со звездочкой», которые помогут тебе в будущем.

<span style='background-color:#F9EDA6'>Жёлтым цветом </span> и значком ⚠️ выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).

Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>образец комментария студента</b></font>
   
Можешь использовать этот комментарий. Просто удали этот текст и вместо него напиши свой 🙂

</div>

# Прекод

# Сборный проект-4

Вам поручено разработать демонстрационную версию поиска изображений по запросу.

Для демонстрационной версии нужно обучить модель, которая получит векторное представление изображения, векторное представление текста, а на выходе выдаст число от 0 до 1 — покажет, насколько текст и картинка подходят друг другу.

### Описание данных

Данные лежат в папке `/datasets/image_search/` или доступны по [ссылке](https://code.s3.yandex.net/datasets/dsplus_integrated_project_4.zip).

В файле `train_dataset.csv` находится информация, необходимая для обучения: имя файла изображения, идентификатор описания и текст описания. Для одной картинки может быть доступно до 5 описаний. Идентификатор описания имеет формат `<имя файла изображения>#<порядковый номер описания>`.

В папке `train_images` содержатся изображения для тренировки модели.

В файле `CrowdAnnotations.tsv` — данные по соответствию изображения и описания, полученные с помощью краудсорсинга. Номера колонок и соответствующий тип данных:

1. Имя файла изображения.
2. Идентификатор описания.
3. Доля людей, подтвердивших, что описание соответствует изображению.
4. Количество человек, подтвердивших, что описание соответствует изображению.
5. Количество человек, подтвердивших, что описание не соответствует изображению.

В файле `ExpertAnnotations.tsv` содержатся данные по соответствию изображения и описания, полученные в результате опроса экспертов. Номера колонок и соответствующий тип данных:

1. Имя файла изображения.
2. Идентификатор описания.

3, 4, 5 — оценки трёх экспертов.

Эксперты ставят оценки по шкале от 1 до 4, где 1 — изображение и запрос совершенно не соответствуют друг другу, 2 — запрос содержит элементы описания изображения, но в целом запрос тексту не соответствует, 3 — запрос и текст соответствуют с точностью до некоторых деталей, 4 — запрос и текст соответствуют полностью.

В файле `test_queries.csv` находится информация, необходимая для тестирования: идентификатор запроса, текст запроса и релевантное изображение. Для одной картинки может быть доступно до 5 описаний. Идентификатор описания имеет формат `<имя файла изображения>#<порядковый номер описания>`.

В папке `test_images` содержатся изображения для тестирования модели.

## 1. Исследовательский анализ данных

Наш датасет содержит экспертные и краудсорсинговые оценки соответствия текста и изображения.

В файле с экспертными мнениями для каждой пары изображение-текст имеются оценки от трёх специалистов. Для решения задачи вы должны эти оценки агрегировать — превратить в одну. Существует несколько способов агрегации оценок, самый простой — голосование большинства: за какую оценку проголосовала большая часть экспертов (в нашем случае 2 или 3), та оценка и ставится как итоговая. Поскольку число экспертов меньше числа классов, может случиться, что каждый эксперт поставит разные оценки, например: 1, 4, 2. В таком случае данную пару изображение-текст можно исключить из датасета.

Вы можете воспользоваться другим методом агрегации оценок или придумать свой.

В файле с краудсорсинговыми оценками информация расположена в таком порядке:

1. Доля исполнителей, подтвердивших, что текст **соответствует** картинке.
2. Количество исполнителей, подтвердивших, что текст **соответствует** картинке.
3. Количество исполнителей, подтвердивших, что текст **не соответствует** картинке.

После анализа экспертных и краудсорсинговых оценок выберите либо одну из них, либо объедините их в одну по какому-то критерию: например, оценка эксперта принимается с коэффициентом 0.6, а крауда — с коэффициентом 0.4.

Ваша модель должна возвращать на выходе вероятность соответствия изображения тексту, поэтому целевая переменная должна иметь значения от 0 до 1.


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 👍 Да, это хорошая практика - дать описание контекста и проблемы, которую мы решаем. </div>

In [1]:
#!pip install spacy
#!python  spacy download en_core_web_sm
#!pip install spacy[transformers]
#!pip install -q torchvision
#!python -m spacy download en_core_web_lg
#!python -m spacy download en_core_web_sm
#!python -m spacy download en_core_web_sm
!pip install  spacy download en_core_web_sm
#!pip3 install https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-2.2.0/en_core_web_sm-2.2.0.tar.gz

Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: http://packages.polyus.com/repository/pypi-proxy/simple
  Using cached http://packages.polyus.com/repository/pypi-proxy/packages/download/0.3.5/download-0.3.5-py3-none-any.whl (8.8 kB)


ERROR: Could not find a version that satisfies the requirement en_core_web_sm (from versions: none)
ERROR: No matching distribution found for en_core_web_sm


<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Лично меня очень утомляют простыни логов после установки библиотек. Я их все равно никто не читает. Можно добавить флаг -q будет гораздо аккуратноее.
    
    !pip install -q torchvision
    
Кстати, я вижу ты делаешь повторную установку torchvision, что для локальной среды уже лишнее. А во многих облачных средах (например в Google Colab) эта библиотека уже установлена.     
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента</b></font>
   
Выполнено

</div>

In [2]:
#import torchvision.models as models
#import torch
#import torch.nn as nn
import numpy as np
#import tensorflow as tf
import math
import pandas as pd
import random as rd

from PIL import Image
import os

import spacy
import re
import nltk
from nltk.corpus import stopwords as nltk_stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import GroupShuffleSplit

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import MinMaxScaler

#from skorch.callbacks import EpochScoring, EarlyStopping
#from skorch import NeuralNetRegressor
from sklearn.pipeline import Pipeline
from tqdm import tqdm, tqdm_notebook
#import torch
#import torch.nn as nn
from math import ceil

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично, все нужные библиотеки импортированы в начале ноутбука.Это хорошая практика.</div>

Импортируем данные.

In [3]:
train_dataset = pd.read_csv('to_upload/train_dataset.csv')
CrowdAnnotations = pd.read_csv('to_upload/CrowdAnnotations.tsv', sep='\t', header=None)
ExpertAnnotations = pd.read_csv('to_upload/ExpertAnnotations.tsv', sep='\t', header=None)
test_queries = pd.read_csv('to_upload/test_queries.csv', sep='|', index_col=0)
test_images = pd.read_csv('to_upload/test_images.csv', sep='|')

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера</b>

Данный проект очень хороший пример того, как один и тот же код может запускаться на разных платформах, у разных людей, даже на операционных системах. Поэтому лучше заранее позаботиться о том, чтобы при переходе с одного компьютера на другой нужно было вносить как можно меньше изменений. Самый простой способ - вынести путь к данным в константу, которую определлить в самом начале ноутбука и все пути определять с этой константой.
    
Также для объединения путей лучше испльзовать os.path.join, т.к. этот метод соединяет пути с учетом операционной системы (в linux и windows для путей используются знаки / и \)
    
Если ты используешь colab или другой облачный сервис, то датасет в виртуальную машину можно загрузить вот так
    
    
    # загрузка датасета 

    !wget -q https://code.s3.yandex.net/datasets/dsplus_integrated_project_4.zip
    !unzip -q dsplus_integrated_project_4.zip

А вот это образец универсального кода для загрузки данных    
    
    DATA_PATH = твой путь    
    
    df_train = pd.read_csv(os.path.join(DATA_PATH, 'train_dataset.csv'))
    ....

 
    
    


</div>

In [4]:
train_dataset.sample(10)

image                     query_id  \
3240  3220161734_77f42734b9.jpg  3197981073_3156963446.jpg#2   
4466   486917990_72bd4069af.jpg   485245061_5a5de43e20.jpg#2   
1430  3139160252_75109e9e05.jpg  2788945468_74a9618cfa.jpg#2   
2010   211295363_49010ca38d.jpg  1472230829_803818a383.jpg#2   
1146  1584315962_5b0b45d02d.jpg  3359530430_249f51972c.jpg#2   
2332  3017521547_f5ef8848e3.jpg  1446053356_a924b4893f.jpg#2   
2184  1425069308_488e5fcf9d.jpg  2251747182_6b67a3ab8b.jpg#2   
2127    54501196_a9ac9d66f2.jpg   533979933_a95b03323b.jpg#2   
1957  1490670858_e122df2560.jpg  2612488996_9450de0e54.jpg#2   
1811  3290105461_7590f23371.jpg   293881927_ac62900fd4.jpg#2   

                                             query_text  
3240  A woman sings into a microphone while playing ...  
4466                         A man kneels in the snow .  
1430  a middle aged man is standing next to a white ...  
2010                            Rafting boat on river .  
1146  People are watching a snowboarder perform a tr...  
2332  Two girls in soccer strips are playing on a sp...  
2184  A young boy , playing basketball , is getting ...  
2127  A boy is on the beach with a paddle in his han...  
1957  A homeless man stands with a sign next to oran...  
1811                            The dog is on the fence

In [5]:
train_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5822 entries, 0 to 5821
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   image       5822 non-null   object
 1   query_id    5822 non-null   object
 2   query_text  5822 non-null   object
dtypes: object(3)
memory usage: 136.6+ KB


In [6]:
train_dataset.duplicated().sum()

0

В тренировочном датасете 5822 строк, или в среднем 5,8 описаний  на каждую из 1000 изображений.

In [7]:
train_dataset.describe()

image                     query_id  \
count                        5822                         5822   
unique                       1000                          977   
top     3364151356_eecd07a23e.jpg  2600867924_cd502fc911.jpg#2   
freq                           10                           34   

                                               query_text  
count                                                5822  
unique                                                977  
top     Two dogs , one brown and white and one black a...  
freq                                                   34

При этом уникальных запросов  977

In [8]:
CrowdAnnotations = pd.read_csv('to_upload/ExpertAnnotations.tsv',  sep='\t', header=None, names=['image', 'query_id', 'percentage_confirmed', 'quantity_confirmed', 'quantity_disproved'])
CrowdAnnotations.head(10)

image                     query_id  \
0  1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2   
1  1056338697_4f7d7ce270.jpg  2718495608_d8533e3ac5.jpg#2   
2  1056338697_4f7d7ce270.jpg  3181701312_70a379ab6e.jpg#2   
3  1056338697_4f7d7ce270.jpg  3207358897_bfa61fa3c6.jpg#2   
4  1056338697_4f7d7ce270.jpg  3286822339_5535af6b93.jpg#2   
5  1056338697_4f7d7ce270.jpg  3360930596_1e75164ce6.jpg#2   
6  1056338697_4f7d7ce270.jpg  3545652636_0746537307.jpg#2   
7  1056338697_4f7d7ce270.jpg   434792818_56375e203f.jpg#2   
8   106490881_5a2dd9b7bd.jpg  1425069308_488e5fcf9d.jpg#2   
9   106490881_5a2dd9b7bd.jpg  1714316707_8bbaa2a2ba.jpg#2   

   percentage_confirmed  quantity_confirmed  quantity_disproved  
0                     1                   1                   1  
1                     1                   1                   2  
2                     1                   1                   2  
3                     1                   2                   2  
4                     1                   1                   2  
5                     1                   1                   1  
6                     1                   1                   1  
7                     1                   1                   2  
8                     1                   1                   1  
9                     2                   2                   2

In [9]:
CrowdAnnotations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5822 entries, 0 to 5821
Data columns (total 5 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   image                 5822 non-null   object
 1   query_id              5822 non-null   object
 2   percentage_confirmed  5822 non-null   int64 
 3   quantity_confirmed    5822 non-null   int64 
 4   quantity_disproved    5822 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 227.5+ KB


In [10]:
CrowdAnnotations.duplicated().sum()

0

In [11]:
CrowdAnnotations.describe()

percentage_confirmed  quantity_confirmed  quantity_disproved
count           5822.000000         5822.000000         5822.000000
mean               1.436620            1.624356            1.881999
std                0.787084            0.856222            0.904087
min                1.000000            1.000000            1.000000
25%                1.000000            1.000000            1.000000
50%                1.000000            1.000000            2.000000
75%                2.000000            2.000000            2.000000
max                4.000000            4.000000            4.000000

Почти 7% из опрошенных подтверждают, что описание соответствует изображению. При этом, имеется максимальное значение — 1, что означает, что есть изображения с описанием, релевантность которого подвердили абсолютно все "краудсорсеры".

In [12]:
ExpertAnnotations = pd.read_csv('to_upload/ExpertAnnotations.tsv', sep='\t', header=None, names=['image', 'query_id', 'expert_1', 'expert_2', 'expert_3'])
ExpertAnnotations.head(10)

image                     query_id  expert_1  expert_2  \
0  1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2         1         1   
1  1056338697_4f7d7ce270.jpg  2718495608_d8533e3ac5.jpg#2         1         1   
2  1056338697_4f7d7ce270.jpg  3181701312_70a379ab6e.jpg#2         1         1   
3  1056338697_4f7d7ce270.jpg  3207358897_bfa61fa3c6.jpg#2         1         2   
4  1056338697_4f7d7ce270.jpg  3286822339_5535af6b93.jpg#2         1         1   
5  1056338697_4f7d7ce270.jpg  3360930596_1e75164ce6.jpg#2         1         1   
6  1056338697_4f7d7ce270.jpg  3545652636_0746537307.jpg#2         1         1   
7  1056338697_4f7d7ce270.jpg   434792818_56375e203f.jpg#2         1         1   
8   106490881_5a2dd9b7bd.jpg  1425069308_488e5fcf9d.jpg#2         1         1   
9   106490881_5a2dd9b7bd.jpg  1714316707_8bbaa2a2ba.jpg#2         2         2   

   expert_3  
0         1  
1         2  
2         2  
3         2  
4         2  
5         1  
6         1  
7         2  
8         1  
9         2

In [13]:
ExpertAnnotations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5822 entries, 0 to 5821
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   image     5822 non-null   object
 1   query_id  5822 non-null   object
 2   expert_1  5822 non-null   int64 
 3   expert_2  5822 non-null   int64 
 4   expert_3  5822 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 227.5+ KB


In [14]:
ExpertAnnotations.duplicated().sum()

0

In [15]:
ExpertAnnotations.describe()

expert_1     expert_2     expert_3
count  5822.000000  5822.000000  5822.000000
mean      1.436620     1.624356     1.881999
std       0.787084     0.856222     0.904087
min       1.000000     1.000000     1.000000
25%       1.000000     1.000000     1.000000
50%       1.000000     1.000000     2.000000
75%       2.000000     2.000000     2.000000
max       4.000000     4.000000     4.000000

Третий эксперт оценивает сходство запроса и картинки выше остальных экспертов (среднее 1,88 против 1,4 и 1,6) 

In [16]:
test_queries = pd.read_csv('to_upload/test_queries.csv', sep='|', index_col=0)
test_queries.head(10)

query_id  \
0  1177994172_10d143cb8d.jpg#0   
1  1177994172_10d143cb8d.jpg#1   
2  1177994172_10d143cb8d.jpg#2   
3  1177994172_10d143cb8d.jpg#3   
4  1177994172_10d143cb8d.jpg#4   
5  1232148178_4f45cc3284.jpg#0   
6  1232148178_4f45cc3284.jpg#1   
7  1232148178_4f45cc3284.jpg#2   
8  1232148178_4f45cc3284.jpg#3   
9  1232148178_4f45cc3284.jpg#4   

                                          query_text  \
0  Two blonde boys , one in a camouflage shirt an...   
1  Two boys are squirting water guns at each other .   
2            Two boys spraying each other with water   
3  Two children wearing jeans squirt water at eac...   
4  Two young boys are squirting water at each oth...   
5                    A baby girl playing at a park .   
6  A closeup of a child on a playground with adul...   
7  A young boy poses for a picture in front of a ...   
8  A young girl is smiling in front of the camera...   
9  There is a little blond hair girl with a green...   

                       image  
0  1177994172_10d143cb8d.jpg  
1  1177994172_10d143cb8d.jpg  
2  1177994172_10d143cb8d.jpg  
3  1177994172_10d143cb8d.jpg  
4  1177994172_10d143cb8d.jpg  
5  1232148178_4f45cc3284.jpg  
6  1232148178_4f45cc3284.jpg  
7  1232148178_4f45cc3284.jpg  
8  1232148178_4f45cc3284.jpg  
9  1232148178_4f45cc3284.jpg

In [17]:
test_queries.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 0 to 499
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   query_id    500 non-null    object
 1   query_text  500 non-null    object
 2   image       500 non-null    object
dtypes: object(3)
memory usage: 15.6+ KB


In [18]:
test_queries.duplicated().sum()

0

In [19]:
test_queries.describe()

query_id  \
count                           500   
unique                          500   
top     1177994172_10d143cb8d.jpg#0   
freq                              1   

                                               query_text  \
count                                                 500   
unique                                                500   
top     Two blonde boys , one in a camouflage shirt an...   
freq                                                    1   

                            image  
count                         500  
unique                        100  
top     1177994172_10d143cb8d.jpg  
freq                            5

Видим 100 изображений, по 5 запросов на каждое.

### ВЫВОДЫ

 * Проведена загрузка данных и созданы переменные для них.
 * Обнаружено, что в работе на краудсорсинге соответствие текстового запроса изображению отмечается менее, чем в 7% случаев.
 * Эксперты в целом также невысоко оценивают соответствие текста изображению, их средние оценки колеблются в диапазоне 1,4-1,9 баллов из 4 и можно сделать вывод что запрос и изображение непохожи.
 * Экспертных оценок у нас 5822 запросов × 3 эксперта, а краудсорсинговых 47830.

### !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!_______________!!!!!!!!!!!!!!!!!!!!!_____

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Ты познакомился с данными. Пожалуйста, напиши выводы что ты в них увидел.

</div>


<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента</b></font>
   
Дополнено выводами

</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
    
Прежде чем приступать к чистке советую тебе провести предобработку. Общая последовательность шагов следующая:
    
 - вначале мы знакомимся с данными и мысленно составляем план, как мы будем "жонглировать" доступными данными, чтобы получить итоговый датафрейм, на котором будем учить модели
 - в рамках реализации этого плана получаем единый датафрейм используя доступные три.
    
В следующем комментарии подсвечу тебе возможные стратегии использования оценок с плюсами и минусами    
    
    
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
    
В этом проекте есть несколько возможных стратегий работы с обучающими и краудсорс оценками,  вот эти стратегии с плюсами и минусами:
   
    
 - использовать соединение через outer, в этом случае, при корректной обработке пропусков мы можем получить более 50К размеченых пар. К плюсам данного подхода можно отнести большое количество данных. Минус - в основном это будут данные, размеченные краудсорсерами, а там качество раметки ниже.
 - использовать только экспертные оценки. плюс - высокое качество данных (разметка имеет шкалу). Минус - данных меньше.  Поясню по поводу качества разметки на примере: если на изображении одна собака, а в описании "две собаки бегут по берегу", то эксперты поставять соответствие 0,6-0,7 (переводя в шкалу 0-1), а краудсорсеры поставят 0.  Но на экспертных оценках модель хотя-бы научится находить собак, а на краудсор оценках ничему не начится.
    
 - использование соединения через left.  Данный подход очевидно проигрывает второму варианту: данных столько же, нужно дополнительно возиться с объединением, но качество разметки снижается.. Но это соответсвует предложениям авторов проекта, поэтому такой подход имеет право на жизнь
    
 - объединение через inner. Тут все плохо - и данных ОЧЕНЬ мало (в два раза меньше, чем использовать только экспертные оценки) и качество у них "подпорчено". И навыков по обработке пропусков не получаем.....
 
</div>

## 2. Проверка данных

В некоторых странах, где работает ваша компания, действуют ограничения по обработке изображений: поисковым сервисам и сервисам, предоставляющим возможность поиска, запрещено без разрешения родителей или законных представителей предоставлять любую информацию, в том числе, но не исключительно тексты, изображения, видео и аудио, содержащие описание, изображение или запись голоса детей. Ребёнком считается любой человек, не достигший 16 лет.

В вашем сервисе строго следуют законам стран, в которых работают. Поэтому при попытке посмотреть изображения, запрещённые законодательством, вместо картинок показывается дисклеймер:

> This image is unavailable in your country in compliance with local laws
>

Однако у вас в PoC нет возможности воспользоваться данным функционалом. Поэтому все изображения, которые нарушают данный закон, нужно удалить из обучающей выборки.

In [20]:
# Удалим пары «изображение-запрос», где от всех трёх экспертов получены разные оценки
unique_votes = ExpertAnnotations.loc[:,['expert_1', 'expert_2', 'expert_3']].apply(pd.Series.nunique, axis=1)
mask = unique_votes <= 2
ExpertAnnotations = ExpertAnnotations[mask]

In [21]:
#Сосчитаем количество голосов каждого эксперта за каждую из оценок
votes = ExpertAnnotations.apply(pd.Series.value_counts, axis=1)

# Найдем оценки, за которые проголосовала бОльшая часть экспертов
majority = votes.idxmax(axis=1)

# Создадим новый столбец для этих оценок
ExpertAnnotations['majority'] = majority
ExpertAnnotations

image                     query_id  expert_1  \
0     1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2         1   
1     1056338697_4f7d7ce270.jpg  2718495608_d8533e3ac5.jpg#2         1   
2     1056338697_4f7d7ce270.jpg  3181701312_70a379ab6e.jpg#2         1   
3     1056338697_4f7d7ce270.jpg  3207358897_bfa61fa3c6.jpg#2         1   
4     1056338697_4f7d7ce270.jpg  3286822339_5535af6b93.jpg#2         1   
...                         ...                          ...       ...   
5817   997722733_0cb5439472.jpg  2981702521_2459f2c1c4.jpg#2         1   
5818   997722733_0cb5439472.jpg  2985679744_75a7102aab.jpg#2         1   
5819   997722733_0cb5439472.jpg  3150742439_b8a352e1e0.jpg#2         1   
5820   997722733_0cb5439472.jpg  3375070563_3c290a7991.jpg#2         1   
5821   997722733_0cb5439472.jpg   486917990_72bd4069af.jpg#2         3   

      expert_2  expert_3  majority  
0            1         1         1  
1            1         2         1  
2            1         2         1  
3            2         2         2  
4            1         2         1  
...        ...       ...       ...  
5817         1         1         1  
5818         1         1         1  
5819         1         2         1  
5820         1         1         1  
5821         3         3         3  

[5696 rows x 6 columns]

In [22]:
# Объединяем таблицы
df = pd.merge(ExpertAnnotations, CrowdAnnotations, on=['image', 'query_id'], how='left').drop(['expert_1', 'expert_2', 'expert_3'], axis=1)

In [23]:
#функция для агрегирования оценок экспертов и крауда
def avg_note(row):
    majority = row.loc['majority']
    percentage_confirmed = row.loc['percentage_confirmed']
    #проверяем наличие оценки от крауда в каждой строоке
    if math.isnan(percentage_confirmed) :
        note = (majority-1)/3
    else:
        note = (majority-1)/3*0.7 + percentage_confirmed*0.3
    return note

df['avg_note']=df.apply(avg_note, axis=1)
df = df.drop(['majority', 'percentage_confirmed', 'quantity_confirmed', 'quantity_disproved'], axis=1)
df

image                     query_id  avg_note
0     1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2  0.300000
1     1056338697_4f7d7ce270.jpg  2718495608_d8533e3ac5.jpg#2  0.300000
2     1056338697_4f7d7ce270.jpg  3181701312_70a379ab6e.jpg#2  0.300000
3     1056338697_4f7d7ce270.jpg  3207358897_bfa61fa3c6.jpg#2  0.533333
4     1056338697_4f7d7ce270.jpg  3286822339_5535af6b93.jpg#2  0.300000
...                         ...                          ...       ...
5691   997722733_0cb5439472.jpg  2981702521_2459f2c1c4.jpg#2  0.300000
5692   997722733_0cb5439472.jpg  2985679744_75a7102aab.jpg#2  0.300000
5693   997722733_0cb5439472.jpg  3150742439_b8a352e1e0.jpg#2  0.300000
5694   997722733_0cb5439472.jpg  3375070563_3c290a7991.jpg#2  0.300000
5695   997722733_0cb5439472.jpg   486917990_72bd4069af.jpg#2  1.366667

[5696 rows x 3 columns]

In [24]:
df.describe()

avg_note
count  5696.000000
mean      0.575193
std       0.426619
min       0.300000
25%       0.300000
50%       0.300000
75%       0.833333
max       1.900000

In [30]:
train_dataset['query_text'] = train_dataset['query_text'].str.lower()

In [31]:
disabled_pipes = ["parser",  "ner"]
nlp = spacy.load("en_core_web_sm", disable=disabled_pipes)

OSError: [E050] Can't find model 'en_core_web_sm'. It doesn't seem to be a Python package or a valid path to a data directory.

In [27]:
def clear_text(text):
    processed_text = re.sub(r'[^A-Za-z ]', ' ', text)
    return ' '.join(processed_text.split())

In [28]:
tqdm.pandas()
train_dataset['query_text'] = train_dataset['query_text'].progress_apply(lambda x: clear_text(x))

100%|███████████████████████████████████████████████████████████████████████████| 5822/5822 [00:00<00:00, 64216.88it/s]


In [29]:
train_dataset['query_text'] = [ " ".join([i.lemma_ for i in doc]) for doc in tqdm(nlp.pipe(train_dataset['query_text'].values, disable=['ner', 'parser']), total=train_dataset.shape[0])]

NameError: name 'nlp' is not defined

In [ ]:
def lemmatize(texts):
    lemmatized_texts = []
    for doc in tqdm(nlp.pipe(texts, disable=['ner', 'parser']), total=len(texts)):
        lemmatized_texts.append(" ".join([i.lemma_ for i in doc]))
    return lemmatized_texts

In [ ]:
stop_kids = ['child', 'children', 'baby', 'babies', 'girl', 'boy', 'girls', 'boys', 'kid', 'kids', 'juvenile',
             'minor', 'minors', 'teen', 'teens', 'teenager', 'teenagers', 'underage', 'infant', 'infants', 'toddler', 'toddlers']

In [ ]:
def check_kids(row):
    text = row.loc['query_text']
    for word in text.split():
        if word in stop_kids:
            return True
    return False

In [ ]:
# (добавляем столбец) создаем список/таблицу с признаком наличия "запрещенных" слов в тексте
df_train['kids_list'] = df_train.apply(check_kids, axis=1)

In [ ]:
df_train.head(30)

In [ ]:
# Фильтрация строк, где kids_list = True
kids_df = df_train[df_train['kids_list'] == True]
# Составление списка query_id
query_id_kids = kids_df['query_id']

In [ ]:
query_id_kids

In [ ]:
Удаляем последние два символа для получения списка названий файлов.

In [ ]:
query_id_kidslist = [query_id[:-2] for query_id in query_id_kids]

In [ ]:
len(query_id_kidslist)

In [ ]:
Удаляем дубликаты названий

In [ ]:
modified_query_id_list = list(set(query_id_kidslist))

In [ ]:
df_train_nokids = df_train[~df_train['image'].isin(modified_query_id_list)]

In [ ]:
df_train_nokids = df_train_nokids.reset_index(drop=True)

In [ ]:
df_train_nokids

Таким образом, мы выполнили фильтрацию тренировочного датасета от изображений с детьми. Осталось 4088 строк из 5818, удалено 1730 строк.

Поскольку финальная модель должна обучаться как на текстовых описаниях и оценках, так и на самих изображениях, соединим df, где есть посчитанная по формуле оценка, и df_train_nokids, где есть текст запроса, и при этом нет изображений с детьми. Джойн будем производить по паре значений 'image' и 'query_id'.

In [ ]:
train_df = pd.merge(df, df_train_nokids, on=['image', 'query_id'], how='inner')
train_df.info()

### !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

In [ ]:
display(ExpertAnnotations.head())
display(CrowdAnnotations.head())
#train_dataset_clear.head()

ExpertAnnotations[6] = (ExpertAnnotations[5]-1)/3
train_data = ExpertAnnotations.merge(CrowdAnnotations, left_on=[0, 1], right_on=[0, 1])[[0, 1, 6, '2_y']]
train_data['res'] = train_data[6]*0.6 + train_data['2_y']*0.4
train_data

In [ ]:
train_dataset_clear = train_dataset_clear.merge(train_data, left_on=['image', 'query_id'], right_on=[0, 1])[train_dataset_clear.columns]
train_data = train_dataset_clear.merge(train_data, left_on=['image', 'query_id'], right_on=[0, 1])[train_data.columns]

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Ты использовал соединение через inner и как следствие потерял слишком много данных... Пожалуйста выбери любую другую страгетию (в комментарии выше я подробно расписал тебе разные варианты..</div>


## 3. Векторизация изображений

Перейдём к векторизации изображений.

Самый примитивный способ — прочесть изображение и превратить полученную матрицу в вектор. Такой способ нам не подходит: длина векторов может быть сильно разной, так как размеры изображений разные. Поэтому стоит обратиться к свёрточным сетям: они позволяют "выделить" главные компоненты изображений. Как это сделать? Нужно выбрать какую-либо архитектуру, например ResNet-18, посмотреть на слои и исключить полносвязные слои, которые отвечают за конечное предсказание. При этом можно загрузить модель данной архитектуры, предварительно натренированную на датасете ImageNet.

Возьмем архитектуру resnet50 со стандартрными весами (как раз полученными на датасете resnet50 ImageNet). Уберем последние два слоя, чтобы получить на выходе вектор. На примере одного из изображений посмотрим, какой вектор получается.

In [ ]:
resnet = models.resnet50(pretrained=True)

for param in resnet.parameters():
    param.requires_grad_(False)

print(list(resnet.children()))

modules = list(resnet.children())[:-1]
resnet = nn.Sequential(*modules)

resnet.eval()

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Ух, какая "портянка". Ты точно хочешь выводить все слои модели на несколько экранов?</div>

In [ ]:
from torchvision import transforms
norm = transforms.Normalize(
    mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])

preprocess = transforms.Compose([
    transforms.Resize(150),
    transforms.CenterCrop(140),
    transforms.ToTensor(),
    norm,
])

In [ ]:
img = Image.open("to_upload/train_images/1056338697_4f7d7ce270.jpg").convert('RGB')

In [ ]:
image_tensor = preprocess(img)
output_tensor = resnet(image_tensor.unsqueeze(0)).flatten()
output_tensor.size()
#output_tensor

Размер выходного вектора без пуллинга был слишком велик, поэтому был уменьшен с помощью пуллинга, чтобы сократить затраты памяти.

Превратим все изображения в датасете в вектора с помощью этой модели. Для этого напишем функцию.

In [ ]:
def vect_gen_test(string):
    img = Image.open(path+string[0]).convert('RGB')
    image_tensor = preprocess(img)
    output_tensor = resnet(image_tensor.unsqueeze(0)).flatten()
    output_tensor.size()
    return output_tensor.numpy()

def vect_gen_train(string):
    img = Image.open(path+string['image']).convert('RGB')
    image_tensor = preprocess(img)
    output_tensor = resnet(image_tensor.unsqueeze(0)).flatten()
    output_tensor.size()
    return output_tensor.numpy()

In [ ]:
path="to_upload/train_images/"
train_img_vec_arr = np.array(train_data.apply(vect_gen_test, axis=1))

#path="to_upload/test_images/"
#test_img_vec_arr = np.array(test_queries.apply(vect_gen_train, axis=1))

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 Отлично, вектра для изображений получены👍 </div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Есть технический момент, связанный с оптимальнойстью кода. Конечно, сейчас у тебя слишком маленький датасет и это возможно не критично, но если ты будешь работать с датасетом крупнее, это может стать проблемой. Вообще у нас должно получиться меньше 1000 уникальных изображений (убираем детские). Но в датасете может быть до 40 тысяч записей (в зависимости от выбранной сттарегии), а значит каждое изображение векторизуется по нескольку раз. Чтобы победить это  нужно сильно переписывать код. Но можно использовать декоратор @lru_cahe (попробуй загуглить про него сама, информации очень много). 
    
Ну и конечно же включение GPU тоже может дать вклад в ускорение векторизации.
    

    from functools import lru_cache
    from tqdm.notebook import tqdm
    tqdm.pandas()    

    @lru_cache(1000)
    def pic_vector(img_name):
        img = Image.open(path+img_name).convert('RGB')
        image_tensor = preprocess(img)
        output_tensor = resnet(image_tensor.unsqueeze(0).to(device)).flatten().cpu()
        output_tensor.size()
        return output_tensor.numpy()    
    
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    resnet = resnet.to(device)    

    df_train_vectors = df_train['image'].progress_apply(pic_vector)

</div>

Таким образом мы получили массив векторов в порядке, в котором изображения находятся в train_data и test_images

## 4. Векторизация текстов

Следующий этап — векторизация текстов. Вы можете поэкспериментировать с несколькими способами векторизации текстов:

- tf-idf
- word2vec
- \*трансформеры (например Bert)

\* — если вы изучали трансформеры в спринте Машинное обучение для текстов.


Воспользуемся tf-idf

In [ ]:
nltk.download('stopwords')
stopwords = set(nltk_stopwords.words('english'))

In [ ]:
corpus = train_dataset_clear['lem_query_text']

count_tf_idf = TfidfVectorizer(stop_words=list(stopwords))

tf_idf_train = count_tf_idf.fit_transform(corpus)

print("Размер матрицы:", tf_idf_train.shape)

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 👍 Вектора для текста получены с помощью TF-IDF. Здесь конечно есть очень тонкий момент, и в другом проекте это мог бы быть красный комментарий. Ты же помнишь, что векторайзеры мы обучаем на обучающем датасете и применяем к валидационому и тестовому. А сейчас ты сделал вектора с использованием всех частей датасета.  Но для этого проекта это должно быть допустимо, т.к. .утечка,повышающая оценку на валидации врядли что-то сильно  испортит.</div>

Теперь можно векторизируем тексты test_queries

In [ ]:
corpus = test_queries['query_text'].apply(lemma_clear)

In [ ]:
test_queries['lem_query_text'] = corpus

In [ ]:
corpus = test_queries['lem_query_text']

tf_idf_test = count_tf_idf.transform(corpus)

print("Размер матрицы:", tf_idf_test.shape)

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Нет смысла делать векторизацию тестовых запросов заранее, лучше внести этот функционал внутрь функции для тестирования
    
    
</div>

### 5. Объединение векторов
Подготовьте данные для обучения: объедините векторы изображений и векторы текстов с целевой переменной.

Чтобы объединить вектора, переведем их в один формат (np.array) и воспользуемся функцией np.hstack

In [ ]:
train_mass = np.hstack([np.vstack(train_img_vec_arr), tf_idf_train.toarray()])


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 Вектора, полученные из текста и изображения объеденены👍 </div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Здесь напрашивается небольшой вывод, хотя бы с указанием размерностей полученных признаков.  
</div>

## 6. Обучение модели предсказания соответствия

Для обучения разделите датасет на тренировочную и тестовую выборки. Простое случайное разбиение не подходит: нужно исключить попадание изображения и в обучающую, и в тестовую выборки.
Для того чтобы учесть изображения при разбиении, можно воспользоваться классом [GroupShuffleSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GroupShuffleSplit.html) из библиотеки sklearn.model_selection.

Код ниже разбивает датасет на тренировочную и тестовую выборки в пропорции 7:3 так, что строки с одинаковым значением 'group_column' будут содержаться либо в тестовом, либо в тренировочном датасете.

```
from sklearn.model_selection import GroupShuffleSplit
gss = GroupShuffleSplit(n_splits=1, train_size=.7, random_state=42)
train_indices, test_indices = next(gss.split(X=df.drop(columns=['target']), y=df['target'], groups=df['group_column']))
train_df, test_df = df.loc[train_indices], df.loc[test_indices]

```

Какую модель использовать — выберите самостоятельно. Также вам предстоит выбрать метрику качества либо реализовать свою.

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
В этом задании требуется выбрать основную метрику, которую ты будешь оптимизировать. Пожаулуйста выбери метрику и обоснуй свой выбор.  </div>

In [ ]:
gss = GroupShuffleSplit(n_splits=1, train_size=.7, random_state=42)
train_indexes, test_indexes = next(gss.split(X=train_mass, y=train_data['res'], groups=train_dataset_clear['image']))
#train_df, test_df = df.loc[train_indices], df.loc[test_indices]

X_train = train_mass[train_indexes]
X_test = train_mass[test_indexes]
y_train = train_data['res'].loc[train_indexes]
y_test = train_data['res'].loc[test_indexes]

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично, благодаря использованию GroupShuffleSplit ты исключаешь попадание одного изображения в обучающий и валидационный наборы.</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Только давай разберемся с выборками. Финальное тестирование (визуальное) мы делаем на отложенной выборке. Поэтому выборка,которую ты сейчас отделил играет роль валидационной. Есть смысл в её названии использовать слово valid а не test</div>

Для начала создадим и обучим линейную регрессию.

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Мы решаем новую задачу и не знаем, наксколько хороши наши метрики и наши модели. Поэтому советую начать с бейзлайна - константной модели (можешь взять простой DummyRegressor).</div>

In [ ]:
model_lr = LinearRegression(positive=True)
#param_search = {}
#gsearch = GridSearchCV(estimator=model_lr, cv=5, param_grid=param_search, scoring='neg_root_mean_squared_error')
#gsearch.fit(X_train, y_train)
model_lr.fit(X_train, y_train)

In [ ]:
scaler = MinMaxScaler()

print("RMSE кросс-валидации:",
      mean_squared_error(y_test, scaler.fit_transform(model_lr.predict(X_test).reshape(-1, 1)), squared=False)
) #С помощью скейлера приводим результаты к значениям от 0 до 1

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Нельзя применять скаллер к предсказаниям, тем более обучать этот скаллер на этих же предсказаниях.

Да, линейная регрессия сильно переобучается. Но давай честно посмотрим правде в глаза и поймем насколько она плоха здесь.
</div>

Теперь создадим полносвязную нейронную сеть.

In [ ]:
X_train = torch.FloatTensor(np.array(X_train))
X_test = torch.FloatTensor(np.array(X_test))
y_train = torch.FloatTensor(np.array(y_train))
y_test = torch.FloatTensor(np.array(y_test))

In [ ]:
X_train.shape

In [ ]:
class Net(nn.Module):
    def __init__(self, n_in_neurons, n_hidden_neurons_1, n_hidden_neurons_2, n_hidden_neurons_3, n_hidden_neurons_4, n_out_neurons):
            super(Net, self).__init__()

            self.fc1 = nn.Linear(n_in_neurons, n_hidden_neurons_1)
            self.act1 = nn.ReLU()
            #self.do1 = nn.Dropout(p=0.7)
            self.fc2 = nn.Linear(n_hidden_neurons_1, n_hidden_neurons_2)
            self.act2 = nn.ReLU()
            self.fc3 = nn.Linear(n_hidden_neurons_2, n_hidden_neurons_3)
            self.act3 = nn.ReLU()
            self.fc4 = nn.Linear(n_hidden_neurons_3, n_hidden_neurons_4)
            self.act4 = nn.ReLU()
            self.fc5 = nn.Linear(n_hidden_neurons_4, n_out_neurons)
            self.act5 = nn.Sigmoid()

            nn.init.kaiming_uniform_(self.fc1.weight, mode='fan_in', nonlinearity='relu')
            nn.init.kaiming_uniform_(self.fc2.weight, mode='fan_in', nonlinearity='relu')
            nn.init.kaiming_uniform_(self.fc3.weight, mode='fan_in', nonlinearity='relu')
            nn.init.kaiming_uniform_(self.fc4.weight, mode='fan_in', nonlinearity='relu')
            nn.init.kaiming_uniform_(self.fc5.weight, mode='fan_in', nonlinearity='sigmoid')

    def forward(self, x):
        x = self.fc1(x)
        x = self.act1(x)
        #x = self.do1(x)
        x = self.fc2(x)
        x = self.act2(x)
        x = self.fc3(x)
        x = self.act3(x)
        x = self.fc4(x)
        x = self.act4(x)
        x = self.fc5(x)
        x = self.act5(x)

        return x

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Технически все верно. Но учитывая что даже простая модель переобучается я бы озаботился борьбой с переобучением. Возможно начал с более простых архитектур (один - два внутренних слоя, возможно поменьше нейронов).

Что касается регуляризации - стоит попробовать dropout (посмотри p в диапазоне 0,1-0,3), но например добавь слои прореживания после каждого внутреннего слоя.
    
Как альтернативу можно попробовать l1/l2 регуляризацию на веса, в pytorch она реализуется через параметр weight_decay оптимайзера.    

</div>

In [ ]:
n_in_neurons = 2800
n_hidden_neurons_1 = 1200
n_hidden_neurons_2 = 600
n_hidden_neurons_3 = 200
n_hidden_neurons_4 = 10
n_out_neurons = 1



net = Net(n_in_neurons, n_hidden_neurons_1, n_hidden_neurons_2, n_hidden_neurons_3, n_hidden_neurons_4, n_out_neurons)

#net = nn.Sequential(
#    nn.Linear(n_in_neurons, n_hidden_neurons_1),
#    nn.ReLU(),
#    nn.Linear(n_hidden_neurons_1, n_hidden_neurons_2),
#    nn.ReLU(),
#    nn.Linear(n_hidden_neurons_2, n_out_neurons)
#)

optimizer2 = torch.optim.SGD(net.parameters(), lr=1e-3, momentum=0.8)

optimizers = [optimizer2]

loss = nn.MSELoss()

In [ ]:
num_epochs = 900
batch_size = 150
num_batches = ceil(len(X_train)/batch_size)

for optimizer in optimizers:
    optimizer_rmse = []
    print(f'Оптимизатор {optimizer}:')

    for epoch in range(num_epochs):
        net.train()
        order = np.random.permutation(len(X_train))

        for batch_idx in range(num_batches):

            start_index = batch_idx * batch_size
            optimizer.zero_grad()

            batch_indexes = order[start_index : start_index + batch_size]
            X_batch = X_train[batch_indexes]
            y_batch = y_train[batch_indexes]

            preds = net.forward(X_batch.float()).flatten()

            loss_value = loss(preds, y_batch)

            loss_value.backward()

            optimizer.step()

        if epoch % 10 == 0 or epoch == num_epochs - 1:
            with torch.no_grad():
                net.eval()
                test_preds = net.forward(X_test).flatten()
                #print(test_preds.numpy())
                ans = round(float(torch.sqrt(loss(test_preds, y_test))), 2)
                print(f'Метрика RMSE на {epoch} итерации =', ans)
                optimizer_rmse.append(ans)
    print()
    print(f'Лучшая метрика RMSE = {min(optimizer_rmse)} обнаружена на {optimizer_rmse.index(min(optimizer_rmse))} эпохе')
    print(f'Средняя метрика RMSE = {round(np.mean(optimizer_rmse), 2)}')
    print('---------------------------------------------------------------------------------------------', '\n')

## 7. Тестирование модели

Настало время протестировать модель. Для этого получите эмбеддинги для всех тестовых изображений из папки `test_images`, выберите случайные 10 запросов из файла `test_queries.csv` и для каждого запроса выведите наиболее релевантное изображение. Сравните визуально качество поиска.

Создадим масив эмбеддингов. Для этого воспользуемся написанной ранее функицей vect_gen_train

In [ ]:
path="to_upload/test_images/"
test_images_vec = np.vstack(np.array(test_images.apply(vect_gen_train, axis=1)))

In [ ]:
def ten_texts():
    texts = []
    for i in range(0,10):
        texts.append(test_queries.iloc[rd.randint(0, 499)]['query_text'])
    return texts

In [ ]:
def conv_text(text):
    corpus_one = lemma_clear(text)
    corpus = []
    #print(corpus_one)
    #print(is_ban_word(corpus_one))
    if not is_ban_word(corpus_one):
        return None
    else:
        #tf_idf_one = count_tf_idf.transform(corpus)
        for i in range(0,100):
            corpus.append(corpus_one)
        tf_idf_one = count_tf_idf.transform(corpus)
        return(np.hstack([np.vstack(test_images_vec), tf_idf_one.toarray()]))
        #return(tf_idf_one)
#conv_text('A young man is wearing blue goggles')

Теперь можно проводить тестирование. Для этого для каждого из 10 подобраных описаний выведем описание и картинку, которая наберет наибольшее количество "очков"

In [ ]:
for item in ten_texts():
    print(item)
    try:
        mass = conv_text(item)
        res = net.forward(torch.FloatTensor(mass)).flatten().detach().numpy()
        #print(res)
        print(res.argmax())
        #print(mass.shape)
    except:
        print('В запросе присутствует запретное слово!')

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Небольшая формальность - пожалуйста добавь в код функции инфереса, чтобы она кроме изображения и текста  выводила найденную меру близости.

Ну и в соответствии с требованиями пожалуйста покажи найденные фото.
</div>

In [ ]:
image_show(images, labels, scores)

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Функции написаны и со своей задачей справляются. Принимают на вход текст и находят изображение. Отрабатывают контроль "запрещенного контента".

Да, мы столкнулись с тем, что модель выдает одно и то же изображение. Это связано с тем, что модель похоже "застряла" в локальном минимуме. Одна из причин такого поведения - использование векторов полученных с помощью TF-IDF, в которых очень много нулей и редко встречаются числа, отличные от нуля. Поэтому модели крайне сложно найти взаимосвязь между такими разноплановыми признаками и таргетом, она слишком быстро переобучается на какие-то другие признаки (тем более и данных у нас немного).
    
Здесь совет только один - поработать с большим объемом датасета + больше экспериментов с архитетурой модели и попытками обучить её. Может получиться (а может и нет).     
</div>

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Идеи и альтернативные подходы</b>
  Если хочешь получить хорошо работающее решение без обучения, то можешь провести эксперимент с нейронной сетью CLIP от OPEN AI (это не обязательно, только если хочешь добиться хороших результатов в этом проекте).

  Про эту нейронку много где написано, но можешь посмотреть например вот здесь  https://habr.com/ru/post/539312/.

  А для написания  реализации наиболее удобной лично для меня оказалась библиотека sbert. https://www.sbert.net/examples/applications/image-search/README.html  
    
    
<hr>    
    
Также неплохо работает следующий подход без предобученных моделей (к тому же легко реализуемый).
    
    
Вместо того, чтобы подавать на вход модели изображение + текст и получать меру соответствия, поступаем по другому. 
    
На вход модели подаем вектор описания, а на выходе учимся предсказывать вектор изображения. 
    
По поводу метрик тут ситуация следующая. Точно работает следующий подход
    
  - берем метрику mse, но тогда нужно чтобы модель её уменьшала. А значит в обучении используем только пары заведомо сответствующие друг другу

Можно попробовать использовать косинусное расстояние, она позволяет использовать все данные (если есть соответствие то выдает 1, если нет то 0). 
    
Также можно попробовать усложнить модель, например и изображения и тексты отправлять в пространство размерностью 256 и уже в этом пространстве учиться их "сближать".
    
По поводу методов векторизации: я бы использовать resnet18 для изображений (размерность эмбеддинга избражения 512) и БЕРТ для текстов (768). 
    
***
    
А еще я видел удачный проект, где для векторизации изображений использовалось предсказание Resnet в котором не отключался последний слой. Получился вектор предсказаний верояятности что изображение принадлежит к каждому из 1000 классов Imagenet. И в сочетании с подходом "мешок слов" для текстов это может дать неплохой результа. Такая модель находит людей, собак, машины, снег и т.д.  
   
</div>

 Модель выдает одно и то же изображение.

## 8. Выводы

- Jupyter Notebook открыт
- Весь код выполняется без ошибок
- Ячейки с кодом расположены в порядке исполнения
- Исследовательский анализ данных выполнен
- Проверены экспертные оценки и краудсорсинговые оценки
- Из датасета исключены те объекты, которые выходят за рамки юридических ограничений
- Изображения векторизованы
- Текстовые запросы векторизованы
- Данные корректно разбиты на тренировочную и тестовую выборки
- Предложена метрика качества работы модели
- Предложена модель схожести изображений и текстового запроса
- Модель обучена
- По итогам обучения модели сделаны выводы
- Проведено тестирование работы модели
- По итогам тестирования визуально сравнили качество поиска

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Заключительный комментарий</b>

Подведем итоги. Я вижу, что в целом с проектом ты справляешься.
    
Проведено исследование и предобработка данных. Выполнена векторизация изображений и текстовых описаний. Обучены модели для предсказания степени соответствия, а также написана функция для инференса и с её помощью проведено тестирование.
 
    
Что нужно доделать обязательно:
    
 - пожалуйста добавь выводы после знакомства с данными
 - пожалуйста используй альтернативный подход к объединению экспертных и краудсорс оценок (не соединение через inner)
 - нужно обосновать выбор оптимизируемой метрики
 - нельзя применять скаллер к предсказаниям, тем более обучать его на них
 - пожалуйста доработай функцию для инференса
    
Желтые комментарии на твое усмотрение. Но учитывая что наш подход "не взлетел" и нужно хоть что-то показать заказчику, попробуй показать возможный инференс с помщью CLIP. С библиотекой sentence-transformers это делается буквально в 10 строк.    
Жду твою работу на повторное ревью.
</div>